[MTM-63943] New MQTT Service - Trust Anchor based auth (#3596)
SoftwareAG/c8y-docs: content/device-integration/mqtt-service-bundle/implementation.md

#### Authentication and authorization {#authentication-and-authorization}

The MQTT Service supports the following authentication methods:

*   **Username and password**
    The MQTT username must include the tenant ID and username in the format `<tenantID>/<username>`.
*   **Device certificates**
    For secure communication, devices must contain the entire chain of certificates leading to the trusted root certificate, or if only the device certificate is provided, then the immediate issuer certificate must be uploaded to the platform’s truststore. You can do this via [the **Trusted certificates** page in the UI](/device-certificate-authentication/managing-trusted-certificates/) or via [REST](https://{{< domain-c8y >}}/api/core/#tag/Trusted-certificates). Moreover, the devices must contain the server certificate in their truststore. 
    <br/>

>     If the trust anchor (that is, the trusted root or intermediate certificate) used to validate the device certificate is trusted by multiple tenants, the device must also specify the tenant ID in the MQTT username field. This ensures that the platform can correctly identify which tenant the device is attempting to connect to. While multi-tenant trust anchors are not currently supported in {{< product-c8y-iot >}}, this feature may be introduced in the future. If the tenant ID is provided, it must correspond to a tenant that trusts the given certificate, otherwise, the connection will be rejected.

Reviewer: Praise: Nice prose, much better than my lame effort ;-)

Author: Taken a few suggestions from AI :)

>     If the trust anchor (that is, the trusted root or intermediate certificate) used to validate the device certificate is trusted by multiple tenants, the device must also specify the tenant ID in the MQTT username field. This ensures that the platform can correctly identify which tenant the device is attempting to connect to. While multi-tenant trust anchors are not currently supported in {{< product-c8y-iot >}}, this feature may be introduced in the future. If the tenant ID is provided, it must correspond to a tenant that trusts the given certificate, otherwise, the connection will be rejected.

Reviewer 2: ```suggestion
    If the trust anchor (that is, the trusted root or intermediate certificate) used to validate the device certificate is trusted by multiple tenants, the device must also specify the tenant ID in the **MQTT username** field. This ensures that the platform can correctly identify which tenant the device is attempting to connect to. While multi-tenant trust anchors are not currently supported in {{< product-c8y-iot >}}, this feature may be introduced in the future. If the tenant ID is provided, it must correspond to a tenant that trusts the given certificate; otherwise, the connection will be rejected.
```

#### Authentication and authorization {#authentication-and-authorization}

The MQTT Service supports the following authentication methods:

*   **Username and password**
    The MQTT username must include the tenant ID and username in the format `<tenantID>/<username>`.
*   **Device certificates**

>     For secure communication, devices must contain the entire chain of certificates leading to the trusted root certificate, or if only the device certificate is provided, then the immediate issuer certificate must be uploaded to the platform’s truststore. You can do it via [the managing trusted certificates in the UI](/device-certificate-authentication/managing-trusted-certificates/) or via [REST](https://{{< domain-c8y >}}/api/core/#tag/Trusted-certificates). Also, the devices must contain the server certificate in their truststore. 

Reviewer 2: ```suggestion
    For secure communication, devices must contain the entire chain of certificates leading to the trusted root certificate, or if only the device certificate is provided, then the immediate issuer certificate must be uploaded to the platform’s truststore. You can do this via [the **Trusted certificates** page in the UI](/device-certificate-authentication/managing-trusted-certificates/) or via [REST](https://{{< domain-c8y >}}/api/core/#tag/Trusted-certificates). Moreover, the devices must contain the server certificate in their truststore. 
```

    <br/>

>     Optionally, the device can also provide the tenant ID in the MQTT username.

Reviewer: Thought: I think we should be a bit clearer about when the tenant id is required (it's not always optional)
- Required if the trust anchor (root or intermediate) for the device certificate is trusted by more than one tenant, to unambiguously identify the tenant that the client is connecting to. This feature (shared trust anchors) is not currently available in Cumulocity but may be added in future.
- Optional if the trust anchor is only trusted by a single tenant.
- If specified, the tenant id must match one of the tenants that trusts the trust anchor, otherwise the connection will be rejected.

I'm sure you can re-word that in a nicer way...

Author: @scmi-c8y , Please check the changes done based on your suggestion: https://github.com/Cumulocity-IoT/c8y-docs/pull/3596/commits/af3ea5eaac8a14c897592be63d844edb23453517

#### Device (client) certificates {#device-certificates}

Using device certificates with the MQTT Service shares the same requirements as outlined in [Device certificates](/device-certificate-authentication/device-certificates#general-requirements-for-connecting-devices-with-certificates).
<br/>

> When connecting devices to the MQTT Service using certificates, the tenant ID can optionally be included in the username field of the MQTT CONNECT packet. This helps the platform unambiguously identify the tenant associated with the device.

Reviewer: Thought: As above, we could be more precise about how the tenant id is used (maybe repeat the same points, or link back to them)

Author: @scmi-c8y , Please check the changes done based on your suggestion: https://github.com/Cumulocity-IoT/c8y-docs/pull/3596/commits/af3ea5eaac8a14c897592be63d844edb23453517

> If the trust anchor (that is, the trusted root or intermediate certificate) used to validate the device certificate is trusted by multiple tenants, the device must also specify the tenant ID in the MQTT username field. This ensures that the platform can correctly identify which tenant the device is attempting to connect to. For more information, see [Authentication and authorization](#authentication-and-authorization).

Reviewer 2: ```suggestion
If the trust anchor (that is, the trusted root or intermediate certificate) used to validate the device certificate is trusted by multiple tenants, the device must also specify the tenant ID in the **MQTT username** field. This ensures that the platform can correctly identify which tenant the device is attempting to connect to. For more information, see [Authentication and authorization](#authentication-and-authorization).
```

#### Adding and trusting CA certificate

TLS trust anchors in the {{< product-c8y-iot >}} platform are defined per tenant.

> To use device certificates for authentication, the CA or intermediate certificate that signs the device certificates must be uploaded to the platform and added to the tenant’s list of trusted certificates. You can do it via [the managing trusted certificates in the UI](/device-certificate-authentication/managing-trusted-certificates/) or via [REST](https://{{< domain-c8y >}}/api/core/#tag/Trusted-certificates).

Reviewer 2: ```suggestion
To use device certificates for authentication, the CA or intermediate certificate that signs the device certificates must be uploaded to the platform and added to the tenant’s list of trusted certificates. You can do this via [the **Trusted certificates** page in the UI](/device-certificate-authentication/managing-trusted-certificates/) or via [REST](https://{{< domain-c8y >}}/api/core/#tag/Trusted-certificates).
```

Explanation:
- `--cafile cumulocity.com.pem`: This file contains the CA certificate of {{< product-c8y-iot >}}’s MQTT Service broker, used to validate the server's identity.
- `--key client-key.pem` and `--cert client-chain.pem`: These are your client certificate and private key, signed by your trusted CA.

> - `-u t11101`: (Optional) Specifies the MQTT username, which must be your tenant ID if provided. In this example, `t11101` is the tenant ID.

Reviewer: Suggestion:
```suggestion
- `-u t11101`: (Optional) Specifies the MQTT username, which must be your tenant ID, as described above. In this example, `t11101` is the tenant ID.
```

Author: @scmi-c8y , Please check the changes done based on your suggestion: https://github.com/Cumulocity-IoT/c8y-docs/pull/3596/commits/af3ea5eaac8a14c897592be63d844edb23453517
